File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        m = len(s)
        n = len(p)

        sIndex = -1
        ans =[]

        sFreq = [0] * 26
        pFreq = [0] * 26 

        for i in range(n):
            pFreq[ord(p[i]) - ord('a')] +=1
        
        l=0
        r=0

        while r < m:
            sFreq[ord(s[r]) - ord('a')] +=1

            if r-l+1 > n:
                sFreq[ord(s[l])-ord('a')] -=1
                l+=1
            
            if sFreq == pFreq:
                ans.append(l)
            
            r+=1
        
        return ans

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int m = s.length();
        int n = p.length();
        vector<int> ans;
        int sIndex = -1;

        if(n>m){
            return  ans;
        }

        vector<int> sFreq(26,0);
        vector<int> pFreq(26,0);

        for(char &ch : p){
            pFreq[ch - 'a']++;
        }

        int l =0;
        int r =0;

        while(r<m){
            sFreq[s[r]-'a']++;

            if(r-l+1 > n){
                sFreq[s[l]-'a']--;
                l++;
            }
            if(sFreq == pFreq){
                ans.push_back(l);
            }
            r++;
        }

        return ans;
    }
};